File: core/completeness.py
"""
SFC file completeness assessment.

Parses AERMET .SFC output files to determine what percentage of hours
have valid (non-missing) meteorological data. Completeness is assessed
from the final SFC output — not from raw inputs — because validity of
a given hour depends on the combination of surface and upper air values.

Key rules:
  - Calm wind hours (zero wind speed) are NOT counted as missing
  - Missing-value sentinel codes indicate invalid hours
  - Completeness = valid_hours / total_hours × 100
"""

import calendar
import os


# SFC file format: fixed-width, one line per hour
# Key columns (approximate — varies by AERMET version):
#   Cols 1-2:   Year (2-digit)
#   Cols 3-4:   Month
#   Cols 5-6:   Day
#   Cols 7-8:   Julian day (3 digits sometimes)
#   Col  9-10:  Hour
#   ...various met parameters...
#
# The SFC format from AERMET is space-delimited with the following fields:
#   YR MO DY JDY HR  H0  U*  W*  DT/DZ ZICNV ZIMCH  M-O  Z0  BOWEN  ALBEDO
#   WSSPD WDIR  ZREF  HT  TTOP  TTREF  ...
#
# Missing value indicators:
#   -999.0 or -9.0 or -9 for various fields
#   The surface heat flux (H0) being -999.0 is a good indicator of a missing hour

# SFC fields (0-indexed after splitting on whitespace):
# 0=YR 1=MO 2=DY 3=JDAY 4=HR 5=H0(sensible heat flux) 6=U* 7=W* 8=DT/DZ
# 9=ZICNV 10=ZIMCH 11=M-O 12=Z0 13=BOWEN 14=ALBEDO
# 15=WSPD 16=WDIR 17=ZREF 18=HT 19=TTOP 20=TTREF

MISSING_VALUE = -999.0


def parse_sfc_file(sfc_path):
    """
    Parse an AERMET .SFC file into a list of hourly records.

    Returns list of dicts with year, month, day, jday, hour, and key met values.
    """
    records = []
    if not os.path.exists(sfc_path):
        return records

    with open(sfc_path, 'r', encoding='utf-8', errors='replace') as f:
        for line in f:
            line = line.strip()
            if not line:
                continue
            parts = line.split()
            if len(parts) < 16:
                continue
            try:
                yr = int(parts[0])
                mo = int(parts[1])
                dy = int(parts[2])
                jday = int(parts[3])
                hr = int(parts[4])
                h0 = float(parts[5])      # sensible heat flux
                ustar = float(parts[6])   # friction velocity
                wstar = float(parts[7])   # convective velocity
                wspd = float(parts[15])   # wind speed
                wdir = float(parts[16])   # wind direction
            except (ValueError, IndexError):
                continue

            records.append({
                'year': yr,
                'month': mo,
                'day': dy,
                'jday': jday,
                'hour': hr,
                'h0': h0,
                'ustar': ustar,
                'wstar': wstar,
                'wspd': wspd,
                'wdir': wdir,
            })
    return records


def is_hour_valid(record):
    """
    Determine if an SFC hour is valid (not missing).

    An hour is valid if the sensible heat flux (H0) is not the missing
    value sentinel. Calm wind hours (wspd = 0) are explicitly valid.
    """
    h0 = record.get('h0', MISSING_VALUE)
    ustar = record.get('ustar', MISSING_VALUE)

    # Primary check: H0 (sensible heat flux) must not be missing
    if h0 <= MISSING_VALUE + 1:  # Close to -999
        return False

    # Secondary check: friction velocity
    if ustar <= MISSING_VALUE + 1:
        return False

    # Calm winds are valid — do NOT flag them as missing
    # (wspd == 0 is fine)

    return True


def compute_completeness(sfc_path, full_year=None):
    """
    Compute hourly data completeness from an SFC file.

    Args:
        sfc_path: Path to .SFC file
        full_year: If specified, compute against expected hours for that year
                   (handles leap years). Otherwise uses hours found in file.

    Returns dict with:
        total_hours: expected hours (8760 or 8784 for leap)
        valid_hours: hours with real data
        missing_hours: hours without valid data
        completeness_pct: percentage valid
        by_quarter: dict of Q1-Q4 completeness
        by_month: dict of month completeness
    """
    records = parse_sfc_file(sfc_path)

    if not records:
        year = full_year or 2020
        total = 8784 if calendar.isleap(year) else 8760
        return {
            'total_hours': total,
            'valid_hours': 0,
            'missing_hours': total,
            'completeness_pct': 0.0,
            'by_quarter': {f'Q{q}': 0.0 for q in range(1, 5)},
            'by_month': {m: 0.0 for m in range(1, 13)},
        }

    # Determine year from records
    year = records[0]['year']
    # Handle 2-digit year
    if year < 100:
        year += 2000 if year < 50 else 1900

    if full_year:
        year = full_year

    is_leap = calendar.isleap(year)
    total_hours = 8784 if is_leap else 8760

    # Quarter definitions (month ranges)
    quarters = {
        'Q1': (1, 2, 3),
        'Q2': (4, 5, 6),
        'Q3': (7, 8, 9),
        'Q4': (10, 11, 12),
    }

    # Count valid hours
    valid_total = 0
    valid_by_month = {m: 0 for m in range(1, 13)}
    hours_by_month = {m: 0 for m in range(1, 13)}

    # Expected hours per month
    expected_by_month = {}
    for m in range(1, 13):
        days = calendar.monthrange(year, m)[1]
        expected_by_month[m] = days * 24

    for rec in records:
        mo = rec['month']
        if 1 <= mo <= 12:
            hours_by_month[mo] += 1
            if is_hour_valid(rec):
                valid_total += 1
                valid_by_month[mo] += 1

    # Monthly completeness
    by_month = {}
    for m in range(1, 13):
        expected = expected_by_month[m]
        by_month[m] = round(valid_by_month[m] / expected * 100, 1) if expected else 0.0

    # Quarterly completeness
    by_quarter = {}
    for qname, months in quarters.items():
        q_expected = sum(expected_by_month[m] for m in months)
        q_valid = sum(valid_by_month[m] for m in months)
        by_quarter[qname] = round(q_valid / q_expected * 100, 1) if q_expected else 0.0

    completeness_pct = round(valid_total / total_hours * 100, 1) if total_hours else 0.0

    return {
        'total_hours': total_hours,
        'valid_hours': valid_total,
        'missing_hours': total_hours - valid_total,
        'completeness_pct': completeness_pct,
        'by_quarter': by_quarter,
        'by_month': by_month,
    }


def check_completeness(sfc_path, threshold, basis='quarterly', full_year=None):
    """
    Check whether an SFC file meets the completeness threshold.

    Args:
        sfc_path: Path to .SFC file
        threshold: Minimum completeness percentage (e.g., 90)
        basis: 'quarterly' or 'annual'
        full_year: Year for expected-hours calculation

    Returns dict with:
        passes: bool
        completeness: full completeness dict
        failures: list of strings describing what failed
    """
    comp = compute_completeness(sfc_path, full_year)
    failures = []

    if basis == 'quarterly':
        for qname, pct in comp['by_quarter'].items():
            if pct < threshold:
                failures.append(f'{qname}: {pct}% (threshold: {threshold}%)')
    else:
        if comp['completeness_pct'] < threshold:
            failures.append(
                f'Annual: {comp["completeness_pct"]}% (threshold: {threshold}%)')

    return {
        'passes': len(failures) == 0,
        'completeness': comp,
        'failures': failures,
    }

File: aermet-tool/src/completeness.rs
/// SFC file completeness assessment.
///
/// Parses AERMET .SFC output files to determine what percentage of hours
/// have valid meteorological data.
use serde::Serialize;
use std::collections::HashMap;
use std::fs;
use std::path::Path;

const MISSING_VALUE: f64 = -999.0;

#[derive(Debug, Clone, Serialize)]
pub struct CompletenessResult {
    pub total_hours: i32,
    pub valid_hours: i32,
    pub missing_hours: i32,
    pub completeness_pct: f64,
    pub by_quarter: HashMap<String, f64>,
    pub by_month: HashMap<String, f64>,
}

#[derive(Debug, Clone, Serialize)]
pub struct CheckResult {
    pub passes: bool,
    pub completeness: CompletenessResult,
    pub failures: Vec<String>,
}

struct SfcRecord {
    month: i32,
    h0: f64,
    ustar: f64,
}

fn is_leap_year(year: i32) -> bool {
    (year % 4 == 0 && year % 100 != 0) || year % 400 == 0
}

fn days_in_month(year: i32, month: i32) -> i32 {
    match month {
        1 => 31,
        2 => {
            if is_leap_year(year) {
                29
            } else {
                28
            }
        }
        3 => 31,
        4 => 30,
        5 => 31,
        6 => 30,
        7 => 31,
        8 => 31,
        9 => 30,
        10 => 31,
        11 => 30,
        12 => 31,
        _ => 30,
    }
}

fn parse_sfc_file(sfc_path: &Path) -> Vec<SfcRecord> {
    let mut records = Vec::new();
    let text = match fs::read_to_string(sfc_path) {
        Ok(t) => t,
        Err(_) => return records,
    };

    for line in text.lines() {
        let line = line.trim();
        if line.is_empty() {
            continue;
        }
        let parts: Vec<&str> = line.split_whitespace().collect();
        if parts.len() < 16 {
            continue;
        }

        let mo: i32 = match parts[1].parse() {
            Ok(v) => v,
            Err(_) => continue,
        };
        let h0: f64 = match parts[5].parse() {
            Ok(v) => v,
            Err(_) => continue,
        };
        let ustar: f64 = match parts[6].parse() {
            Ok(v) => v,
            Err(_) => continue,
        };

        records.push(SfcRecord {
            month: mo,
            h0,
            ustar,
        });
    }
    records
}

fn is_hour_valid(rec: &SfcRecord) -> bool {
    // H0 (sensible heat flux) must not be missing
    if rec.h0 <= MISSING_VALUE + 1.0 {
        return false;
    }
    // Friction velocity must not be missing
    if rec.ustar <= MISSING_VALUE + 1.0 {
        return false;
    }
    // Calm winds are valid
    true
}

pub fn compute_completeness(sfc_path: &Path, full_year: Option<i32>) -> CompletenessResult {
    let records = parse_sfc_file(sfc_path);

    let year = full_year.unwrap_or(2020);
    let total_hours = if is_leap_year(year) { 8784 } else { 8760 };

    if records.is_empty() {
        let by_quarter: HashMap<String, f64> = (1..=4)
            .map(|q| (format!("Q{q}"), 0.0))
            .collect();
        let by_month: HashMap<String, f64> = (1..=12)
            .map(|m| (m.to_string(), 0.0))
            .collect();
        return CompletenessResult {
            total_hours,
            valid_hours: 0,
            missing_hours: total_hours,
            completeness_pct: 0.0,
            by_quarter,
            by_month,
        };
    }

    // Expected hours per month
    let mut expected_by_month: HashMap<i32, i32> = HashMap::new();
    for m in 1..=12 {
        expected_by_month.insert(m, days_in_month(year, m) * 24);
    }

    let mut valid_total = 0;
    let mut valid_by_month: HashMap<i32, i32> = (1..=12).map(|m| (m, 0)).collect();

    for rec in &records {
        if (1..=12).contains(&rec.month) && is_hour_valid(rec) {
            valid_total += 1;
            *valid_by_month.entry(rec.month).or_insert(0) += 1;
        }
    }

    // Monthly completeness
    let by_month: HashMap<String, f64> = (1..=12)
        .map(|m| {
            let expected = expected_by_month[&m] as f64;
            let valid = valid_by_month[&m] as f64;
            let pct = if expected > 0.0 {
                (valid / expected * 1000.0).round() / 10.0
            } else {
                0.0
            };
            (m.to_string(), pct)
        })
        .collect();

    // Quarterly completeness
    let quarters: [(String, &[i32]); 4] = [
        ("Q1".to_string(), &[1, 2, 3]),
        ("Q2".to_string(), &[4, 5, 6]),
        ("Q3".to_string(), &[7, 8, 9]),
        ("Q4".to_string(), &[10, 11, 12]),
    ];

    let by_quarter: HashMap<String, f64> = quarters
        .iter()
        .map(|(qname, months)| {
            let q_expected: i32 = months.iter().map(|m| expected_by_month[m]).sum();
            let q_valid: i32 = months.iter().map(|m| valid_by_month[m]).sum();
            let pct = if q_expected > 0 {
                (q_valid as f64 / q_expected as f64 * 1000.0).round() / 10.0
            } else {
                0.0
            };
            (qname.clone(), pct)
        })
        .collect();

    let completeness_pct = if total_hours > 0 {
        (valid_total as f64 / total_hours as f64 * 1000.0).round() / 10.0
    } else {
        0.0
    };

    CompletenessResult {
        total_hours,
        valid_hours: valid_total,
        missing_hours: total_hours - valid_total,
        completeness_pct,
        by_quarter,
        by_month,
    }
}

pub fn check_completeness(
    sfc_path: &Path,
    threshold: f64,
    basis: &str,
    full_year: Option<i32>,
) -> CheckResult {
    let comp = compute_completeness(sfc_path, full_year);
    let mut failures = Vec::new();

    if basis == "quarterly" {
        for (qname, pct) in &comp.by_quarter {
            if *pct < threshold {
                failures.push(format!("{qname}: {pct}% (threshold: {threshold}%)"));
            }
        }
    } else if comp.completeness_pct < threshold {
        failures.push(format!(
            "Annual: {}% (threshold: {threshold}%)",
            comp.completeness_pct
        ));
    }

    CheckResult {
        passes: failures.is_empty(),
        completeness: comp,
        failures,
    }
}
